File: src/main/java/com/oreilly/learningsparkexamples/java/logs/LogAnalyzerWindowed.java
package com.oreilly.learningsparkexamples.java.logs;

import com.google.common.collect.Ordering;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;
import scala.Tuple4;

import java.io.Serializable;
import java.util.Comparator;
import java.util.List;

public class LogAnalyzerWindowed implements Serializable {
  private LogStatistics logStatistics;

  public void processAccessLogs(String outDir, JavaDStream<ApacheAccessLog> accessLogsDStream) {
    JavaDStream<ApacheAccessLog> windowDStream = accessLogsDStream.window(
        Flags.getInstance().getWindowLength(),
        Flags.getInstance().getSlideInterval());
    JavaDStream<String> ip = accessLogsDStream.map(
      new Function<ApacheAccessLog, String>() {
        public String call(ApacheAccessLog entry) {
          return entry.getIpAddress();
        }});
    // reduceByWindow
    JavaDStream<Long> requestCountRBW = accessLogsDStream.map(new Function<ApacheAccessLog, Long>() {
        public Long call(ApacheAccessLog entry) {
          return 1L;
        }}).reduceByWindow(new Function2<Long, Long, Long>() {
            public Long call(Long v1, Long v2) {
              return v1+v2;
            }}, new Function2<Long, Long, Long>() {
            public Long call(Long v1, Long v2) {
              return v1-v2;
            }}, Flags.getInstance().getWindowLength(), Flags.getInstance().getSlideInterval());
    requestCountRBW.print();
    // reducebykeyandwindow
    JavaPairDStream<String, Long> ipAddressPairDStream = accessLogsDStream.mapToPair(
      new PairFunction<ApacheAccessLog, String, Long>() {
        public Tuple2<String, Long> call(ApacheAccessLog entry) {
          return new Tuple2(entry.getIpAddress(), 1L);
        }});
    JavaPairDStream<String, Long> ipCountDStream = ipAddressPairDStream.reduceByKeyAndWindow(
      // Adding elements in the new slice
      new Function2<Long, Long, Long>() {
        public Long call(Long v1, Long v2) {
          return v1+v2;
        }},
      // Removing elements from the oldest slice
      new Function2<Long, Long, Long>() {
        public Long call(Long v1, Long v2) {
          return v1-v2;
        }},
      Flags.getInstance().getWindowLength(),
      Flags.getInstance().getSlideInterval());
    ipCountDStream.print();
    // Use countByWindow
    JavaDStream<Long> requestCount = accessLogsDStream.countByWindow(
      Flags.getInstance().getWindowLength(), Flags.getInstance().getSlideInterval());
    JavaPairDStream<String, Long> ipAddressRequestCount = ip.countByValueAndWindow(
      Flags.getInstance().getWindowLength(), Flags.getInstance().getSlideInterval());
    requestCount.print();
    ipAddressRequestCount.print();

    // use a transform for the response code count
    JavaPairDStream<Integer, Long> responseCodeCountTransform = accessLogsDStream.transformToPair(
      new Function<JavaRDD<ApacheAccessLog>, JavaPairRDD<Integer, Long>>() {
        public JavaPairRDD<Integer, Long> call(JavaRDD<ApacheAccessLog> logs) {
          return Functions.responseCodeCount(logs);
        }
      });
    windowDStream.foreachRDD(new Function<JavaRDD<ApacheAccessLog>, Void>() {
        public Void call(JavaRDD<ApacheAccessLog> accessLogs) {
      Tuple4<Long, Long, Long, Long> contentSizeStats =
          Functions.contentSizeStats(accessLogs);

      List<Tuple2<Integer, Long>> responseCodeToCount =
          Functions.responseCodeCount(accessLogs)
          .take(100);

      JavaPairRDD<String, Long> ipAddressCounts =
          Functions.ipAddressCount(accessLogs);
      List<String> ip = Functions.filterIPAddress(ipAddressCounts)
          .take(100);

      Object ordering = Ordering.natural();
      Comparator<Long> cmp = (Comparator<Long>)ordering;
      List<Tuple2<String, Long>> topEndpoints =
          Functions.endpointCount(accessLogs)
        .top(10, new Functions.ValueComparator<String, Long>(cmp));

      logStatistics = new LogStatistics(contentSizeStats, responseCodeToCount,
          ip, topEndpoints);
      return null;
        }});
  }

  public LogStatistics getLogStatistics() {
    return logStatistics;
  }
}

File: src/main/scala/com/oreilly/learningsparkexamples/scala/logs/LogAnalyzerWindowed.scala
package com.oreilly.learningsparkexamples.scala.logs;

import org.apache.spark._
import org.apache.spark.rdd._
import org.apache.spark.SparkContext._
import org.apache.spark.streaming._
import org.apache.spark.streaming.StreamingContext._
import org.apache.spark.streaming.dstream._
import com.oreilly.learningsparkexamples.java.logs.ApacheAccessLog
import org.apache.hadoop.mapred.SequenceFileOutputFormat;
import org.apache.hadoop.io.{ArrayWritable, BooleanWritable, BytesWritable, DoubleWritable, FloatWritable, IntWritable, LongWritable, NullWritable, Text, Writable}

/**
 * Computes various pieces of information on a sliding window form the log input
 */
object LogAnalyzerWindowed {
  def responseCodeCount(accessLogRDD: RDD[ApacheAccessLog]) = {
    accessLogRDD.map(log => (log.getResponseCode(), 1)).reduceByKey((x, y) => x + y)
  }

  def processAccessLogs(accessLogsDStream: DStream[ApacheAccessLog], opts: Config) {
    val ipDStream = accessLogsDStream.map{entry => entry.getIpAddress()}
    val ipAddressRequestCount = ipDStream.countByValueAndWindow(
      opts.getWindowDuration(), opts.getSlideDuration())
    ipAddressRequestCount.saveAsTextFiles(opts.OutputDirectory + "/ipAddressRequestCountsTXT")
    val writableIpAddressRequestCount = ipAddressRequestCount.map{case (ip, count) =>
      (new Text(ip), new LongWritable(count))}
    writableIpAddressRequestCount.saveAsHadoopFiles[SequenceFileOutputFormat[Text, LongWritable]](
      opts.OutputDirectory + "/ipAddressRequestCounts", "pandas")
    val requestCount = accessLogsDStream.countByWindow(opts.getWindowDuration(), opts.getSlideDuration())
    requestCount.print()
    ipAddressRequestCount.print()
    val accessLogsWindow = accessLogsDStream.window(
      opts.getWindowDuration(), opts.getSlideDuration())
    accessLogsWindow.transform(rdd => responseCodeCount(rdd)).print()
    // compute the visit counts for IP address in a window
    val ipPairDStream = accessLogsDStream.map(logEntry => (logEntry.getIpAddress(), 1))
    val ipCountDStream = ipPairDStream.reduceByKeyAndWindow(
      {(x, y) => x + y}, // Adding elements in the new slice
      {(x, y) => x - y}, // Removing elements from the oldest slice
      opts.getWindowDuration(), // Window duration
      opts.getSlideDuration() // slide duration
    )
    ipCountDStream.print()
  }
}
